src/mame/atari/atari400.cpp

```diff
@@ -2095,12 +2229,46 @@ uint8_t a400_state::djoy_b_r()
 	return b;
 }
 
-void a400_state::a600xl_pia_pb_w(uint8_t data)
+// TODO: make sure direction bits are honored for all these portb defs.
+/*
+ * x--- ---- (1) 0x5000-0x57ff self-test ROM enabled if kernel ROM enabled (0) RAM or NOP
+ * ---- --x- (1) 0xa000-0xbfff BASIC ROM enabled (0) disabled
+ * ---- ---x (1) 0xd800-0xffff kernel ROM enable (0) RAM
+ */
+void a800xl_state::pia_portb_w(uint8_t data)
+{
+	if (m_pia->port_b_z_mask() != 0xff)
+	{
+		m_mmu = data;
+	}
+}
+
+// TODO: a600xl_state shouldn't really virtualize this (same as a800xl portb), investigate
+void a600xl_state::pia_portb_w(uint8_t data)
 {
 	m_mmu = data;
 }
 
-void a400_state::a800xl_pia_pb_w(uint8_t data)
+/*
+ * same as a800xl plus:
+ * ---- xx-- LED states (specific for this variant only)
+ */
+void a1200xl_state::pia_portb_w(uint8_t data)
+{
+	const u8 dir = m_pia->port_b_z_mask();
+
+	// BIOS jumps from PC=0xc550, expecting ROM to be still enabled at PC=0xe40c
+	// it expects the dir to work properly.
+	m_mmu = (data & ~dir) | (m_mmu & dir);
```

Comment:
@ajrhacker from my understanding this should be the correct handling, please confirm so that I can backport to all the other portb handlers.